File: lab/c/SIMULATION_POO/produto.c
#include <stdio.h>
#include <string.h>

typedef struct {
    char nome[50];
    float preco;
    int quantidade;
} Produto;

float valor_em_estoque(Produto* p) {
    return p->preco * p->quantidade;
}

void mostrar_produto(Produto* p) {
    if (p->preco == 0)
        printf("SEM VALOR DEFINIDO\n");
    else if (p->quantidade == 0)
        printf("ITEM FORA DE ESTOQUE OU ESGOTADO\n");
    else
        printf("Produto (%s, %d, %.2f) - R$ %.2f\n", p->nome, p->quantidade, p->preco, valor_em_estoque(p));
}

void vender(Produto* p, int qtd) {
    if (qtd >= p->quantidade) {
        printf("Item fora de estoque, por favor pegue outro produto\n");
        return;
    }
    p->quantidade -= qtd;
    printf("Ficou: R$ %.2f\n", valor_em_estoque(p));
    printf("Quantidade em estoque: %d\n", p->quantidade);
}

int main() {
    Produto p;
    strcpy(p.nome, "Chave Micha - Kit");
    p.preco = 45.98;
    p.quantidade = 5;

    mostrar_produto(&p);

    vender(&p, 2);
    vender(&p, 2);
    vender(&p, 1);

    return 0;
}

File: lab/javascript/POO/produto.js
class Produto {
  constructor(nome = "SEM NOME", preco = 0.0, quantidade = 0) {
    this.nome = nome;
    this.preco = preco;
    this.quantidade = quantidade;
  }

  valorEmEstoque() {
    return this.preco * this.quantidade;
  }

  toString() {
    let r;
    if (this.preco === 0) r = "SEM VALOR DEFINIDO";
    else if (this.quantidade === 0) r = "ITEM FORA DE ESTOQUE OU ESGOTADO";
    else r = this.valorEmEstoque().toFixed(2);
    return `Produto (${this.nome}, ${this.quantidade}, ${this.preco}) - R$ ${r}`;
  }
}

class Vendas {
  constructor(produto) {
    this.produto = produto;
  }

  vender(quantidade) {
    if (quantidade >= this.produto.quantidade) {
      console.log("Item fora de estoque, por favor pegue outro produto");
      return;
    }
    this.produto.quantidade -= quantidade;
    console.log(`Ficou: R$ ${this.produto.valorEmEstoque().toFixed(2)}`);
    console.log(`Quantidade em estoque: ${this.produto.quantidade}`);
  }
}


const p = new Produto("Chave Micha - Kit", 45.98, 5);
console.log(p.toString());

const venda1 = new Vendas(p);
venda1.vender(2);
venda1.vender(2);
venda1.vender(1);
